sa/sa.go

```diff
@@ -1312,3 +1313,125 @@ func (ssa *SQLStorageAuthority) UpdateCRLShard(ctx context.Context, req *sapb.Up
 
 	return &emptypb.Empty{}, nil
 }
+
+// PauseIdentifiers pauses a set of identifiers for the provided account. If an
+// identifier is currently paused, this is a no-op. If an identifier was
+// previously paused and unpaused, it will be repaused. All work is accomplished
+// in a transaction to limit possible race conditions.
+func (ssa *SQLStorageAuthority) PauseIdentifiers(ctx context.Context, req *sapb.PauseRequest) (*sapb.PauseIdentifiersResponse, error) {
+	if core.IsAnyNilOrZero(req.RegistrationID, req.Identifiers) {
+		return nil, errIncompleteRequest
+	}
+
+	// Marshal the identifier now that we've crossed the RPC boundary.
+	identifiers, err := newIdentifierModelsFromPB(req.Identifiers)
+	if err != nil {
+		return nil, err
+	}
+
+	response := &sapb.PauseIdentifiersResponse{}
+	_, err = db.WithTransaction(ctx, ssa.dbMap, func(tx db.Executor) (interface{}, error) {
+		for _, identifier := range identifiers {
+			pauseError := func(op string, err error) error {
+				return fmt.Errorf("while %s identifier %s for registration ID %d: %w",
+					op, identifier.Value, req.RegistrationID, err,
+				)
+			}
+
+			var entry pausedModel
+			err := tx.SelectOne(ctx, &entry, `
+			SELECT pausedAt, unpausedAt
+			FROM paused
+			WHERE 
+				registrationID = ? AND 
+				identifierType = ? AND 
+				identifierValue = ?`,
+				req.RegistrationID,
+				identifier.Type,
+				identifier.Value,
+			)
+			if err != nil && !errors.Is(err, sql.ErrNoRows) {
```

Comment:
There's a potential issue here, which is that if we ever refactored these cases or added a new case and forgot the "continue" at the end of it, we'd end up executing more than one case inside the transaction.

One thing we could do to prevent that would be using a switch/case, which guarantees exactly one case will be executed:

```go
switch {
case err != nil && !errors.Is(err, sql.ErrNoRows):
  // error out
case errors.Is(err, sql.ErrNoRows):
  // insert
case entry.UnpausedAt == nil || entry.PausedAt.After(entry.UnpausedAt):
  // already paused
case entry.UnpausedAt.After(entry.PausedAt):
  // update
case default:
  return nil, fmt.Errorf("impossible database state encoutered")
}
```

But maybe that's too wild?

Reply:
Ooh, not too wild at all!